sql/spanner/driver.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package spanner
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"ariga.io/atlas/sql/internal/sqlx"
+	"ariga.io/atlas/sql/migrate"
+	"ariga.io/atlas/sql/schema"
+	"ariga.io/atlas/sql/sqlclient"
+
+	// go-sql-spanner provides a database/sql interface to Cloud Spanner.
+	_ "github.com/googleapis/go-sql-spanner"
+)
+
+type (
+	// Driver represents a Spanner driver for introspecting database schemas,
+	// generating diff between schema elements and apply migrations changes.
+	Driver struct {
+		conn
+		schema.Differ
+		schema.Inspector
+		migrate.PlanApplier
+	}
+
+	// database connection and its information.
+	conn struct {
+		schema.ExecQuerier
+		databaseDialect string
+	}
+)
+
+func init() {
+	sqlclient.Register(
+		"spanner",
+		sqlclient.DriverOpener(Open),
+		sqlclient.RegisterCodec(MarshalHCL, EvalHCL),
+		sqlclient.RegisterURLParser(sqlclient.URLParserFunc(func(u *url.URL) *sqlclient.URL {
+			uc := &sqlclient.URL{URL: u, DSN: strings.TrimPrefix(u.String(), u.Scheme+"://")}
+			return uc
+		})),
+	)
+}
+
+// Open opens a new Spanner driver.
+func Open(db schema.ExecQuerier) (migrate.Driver, error) {
+	var (
+		c   = conn{ExecQuerier: db}
+		ctx = context.Background()
+	)
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
```

Comment:
I added this because the spanner emulator doesn't implement short timeouts and you can end up waiting easily and at the same time we're testing against running spanner instances which can take more than a second for some inspection queries -> I'll lower this to 10s.